### PR TITLE
create feature-classification_dataset

### PR DESCRIPTION
**Pull request by VladislavPatrushev**
create feature-classification_dataset

### CODE REVIEW

File: src/data/datasets/classification/image_classification.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    def __init__(self,
+                 data_folder: str,
+                 path_to_datalist: str,
+                 transform: Optional[Union[BasicTransform, BaseCompose]],
+                 augment: Optional[Union[BasicTransform, BaseCompose]] = None,
+                 input_dtype: str = 'float32',
+                 target_dtype: str = 'long',
+                 target_column: str = 'label',
+                 grayscale: bool = False,
+                 expand_rate: int = 1,
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:27]: Remove support of expand rate

File: src/data/datasets/classification/image_classification.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                 test_mode: Optional[bool] = False,
+                 multilabel: bool = False,
+                 num_classes: int = None,
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:30]: `num_classes` must be a required param

File: src/data/datasets/classification/image_classification.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        else:
+            label = list(map(int, re.findall(r'\d+', label)))
+
+        max_label = max(label)
+
+        if max_label < num_classes:
+            multihot = np.zeros((num_classes,), dtype=bool)
+            multihot[label] = True
+        else:
+            raise Exception(f'Target column contain label: {max_label}, it\'s more than num_classes = {num_classes}')
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:131]: `raise ValueError`

File: src/data/datasets/classification/image_classification.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        image = self._read_image(record)
+        sample = {'input': image, 'index': idx}
+        sample = self.apply_transform(self.augment, sample)
+        sample = self.apply_transform(self.transform, sample)
+        sample['input'] = sample['input'].type(torch.__dict__[self.input_dtype])
+
+        if self.test_mode:
+            return sample
+
+        sample['target'] = record[self.target_column]
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:104]: Add method `process_multiclass` where you will check the class index to fit the range

File: src/data/datasets/classification/image_classification.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        sample['target'] = torch.tensor(sample['target']).type(torch.__dict__[self.target_dtype])
+
+        return sample
+
+    def __len__(self) -> int:
+        csv_len = len(self.csv) * self.expand_rate
+        return csv_len
+
+    @staticmethod
+    def process_multilabel(label: Union[int, str], num_classes: int) -> np.array:
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:119]: Add documentation what formats of input this method supports.  
Make it private

File: src/data/datasets/classification/image_classification.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+        if multilabel:
+            self.__num_classes = num_classes
+            self.__lazy_init_multilabel = lazy_init_multilabel
+            if not self.lazy_init_multilabel and not self.test_mode:
+                process_func = partial(self.process_multilabel, num_classes=self.num_classes)
+                self.csv[self.target_column] = self.csv[self.target_column].apply(process_func)
+
+    @property
+    def target_column(self) -> str:
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:69]: Move all properties to the end of class definition

File: src/data/datasets/classification/image_classification.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            augment: Optional augment to be applied on a sample.
+                This should have the interface of transforms in `albumentations` library.
+            input_dtype: Data type of of the torch tensors related to the image.
+            target_dtype: Data type of of the torch tensors related to the target.
+            target_column: Name of the column that contains target labels.
+            grayscale: If True, image will be read as grayscale otherwise as RGB.
+            expand_rate: A multiplier that shows how many times the dataset will be larger than its real size.
+                Useful for small datasets.
+            test_mode: If True, only image without labels will be returned.
+            multilabel: If True, target labels is being converted to multihot vector for multilabel task.
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:48]: `If True, target labels are being converted to multihot vector for multilabel task.`

File: src/data/datasets/classification/image_classification.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            num_classes: Number of classes for multilabel task(length of multihot target vector).
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:49]: `Number of classes (i.e. maximum class index in the dataset).`

File: tests/data/test_image_classification.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+from src.data.datasets.classification.image_classification import ImageClassificationDataset
+
+
+class TestClassification(unittest.TestCase):
+    def __init__(self, methodName: str = None) -> None:
+        super().__init__(methodName)
+        self.data_folder = '/workdir/'
+        self.path_to_datalist = 'test_classification.csv'
+        self.transform = Compose([Resize(224, 224), ToTensorV2()], p=1.0)
```
Comment (VladVin) [inline tests/data/test_image_classification.py:16]: Add augmentations

File: tests/data/test_image_classification.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+    def test_len_default(self):
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform)
+        self.assertEqual(len(self.ds), 3180)
+
+    def test_expand_rate(self):
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform, expand_rate=2)
+        self.assertEqual(len(self.ds), 6360)
+
+    def test_shape_default(self):
```
Comment (VladVin) [inline tests/data/test_image_classification.py:26]: `test_shape_when_transformed`

File: tests/data/test_image_classification.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform)
+        self.assertEqual(list(self.ds[0]['input'].shape), [3, 224, 224])
+
+    def test_grayscale(self):
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform, grayscale=True)
+        self.assertEqual(list(self.ds[0]['input'].shape), [1, 224, 224])
```
Comment (VladVin) [inline tests/data/test_image_classification.py:32]: Without `list`

File: tests/data/test_image_classification.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        input_dtype = 'float32'
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform,
+                                             input_dtype=input_dtype)
+        self.assertEqual(self.ds[1]['input'].dtype, torch.__dict__[input_dtype])
+
+    def test_input_test_mode_true(self):
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform, test_mode=True)
+        self.assertListEqual([*self.ds[0]], ['input', 'index'])
+
+    def test_input_test_mode_false(self):
```
Comment (VladVin) [inline tests/data/test_image_classification.py:44]: `test_getitem_when_test_mode_false`

File: tests/data/test_image_classification.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    def test_multilabel_true(self):
+        num_classes = 9
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform, multilabel=True,
+                                             num_classes=num_classes)
+        self.assertEqual(len(self.ds[0]['target']), num_classes)
+
+    def test_lazy_init_multilabel(self):
+        num_classes = 9
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform, multilabel=True,
+                                             num_classes=9, lazy_init_multilabel=True)
```
Comment (VladVin) [inline tests/data/test_image_classification.py:57]: Put `num_classes`

File: tests/data/test_image_classification.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+    def test_shape_default(self):
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform)
+        self.assertEqual(list(self.ds[0]['input'].shape), [3, 224, 224])
+
+    def test_grayscale(self):
+        self.ds = ImageClassificationDataset(self.data_folder, self.path_to_datalist, self.transform, grayscale=True)
+        self.assertEqual(list(self.ds[0]['input'].shape), [1, 224, 224])
+
+    def test_input_dtype(self):
```
Comment (VladVin) [inline tests/data/test_image_classification.py:34]: Add test: `test_target_dtype_when_specified`

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/data/datasets/classification/image_classification.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+        if not self.__lazy_init and not self.test_mode:
+            if multilabel:
+                self.csv[self.target_column] = self.csv[self.target_column].apply(self.__process_multilabel)
+            else:
+                self.csv[self.target_column] = self.csv[self.target_column].apply(self.__process_multiclass)
+
+    def __getitem__(self, idx: int) -> dict:
+        record = self.csv.iloc[idx]
+        image_path = record[self.input_column]
```
Comment (VladislavPatrushev) [inline src/data/datasets/classification/image_classification.py:68]: Maybe send idx to __read_image ?

File: src/data/datasets/classification/image_classification.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        sample = self._apply_transform(self.augment, sample)
+        sample = self._apply_transform(self.transform, sample)
+        sample['input'] = sample['input'].type(torch.__dict__[self.input_dtype])
+
+        if self.test_mode:
+            return sample
+
+        sample['target'] = record[self.target_column]
+
+        if self.lazy_init:
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:80]: `self.__lazy_init`

File: src/data/datasets/classification/image_classification.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+            if self.multilabel:
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:81]: `self.__multilabel`

File: src/data/datasets/classification/image_classification.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+            augment: Optional augment to be applied on a sample.
+                This should have the interface of transforms in `albumentations` library.
+            input_dtype: Data type of of the torch tensors related to the image.
+            input_column: Name of the column that contains paths to images.
+            target_dtype: Data type of of the torch tensors related to the target.
+            target_column: Name of the column that contains target labels.
+            grayscale: If True, image will be read as grayscale otherwise as RGB.
+            test_mode: If True, only image without labels will be returned.
+            multilabel: If True, target labels are being converted to multihot vector for multilabel task.
+                        If False, dataset prepare target for multiclass classification.
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:48]: prepares targets

File: src/data/datasets/classification/image_classification.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+                 csv_path: str,
+                 num_classes: int,
+                 transform: Optional[Union[BasicTransform, BaseCompose]],
+                 augment: Optional[Union[BasicTransform, BaseCompose]] = None,
+                 input_dtype: str = 'float32',
+                 target_dtype: str = 'long',
+                 input_column: str = 'image_path',
+                 target_column: str = 'label',
+                 grayscale: bool = False,
+                 test_mode: Optional[bool] = False,
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:28]: Remove `Optional`

File: src/data/datasets/classification/image_classification.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+
+        Returns:
+            Verified class index.
+
+        Raises:
+            ValueError: If class index is out of range.
+        """
+        class_idx = class_idx if isinstance(class_idx, int) else int(class_idx)
+        if class_idx >= self.num_classes:
+            raise ValueError(f'Target column contain class index: {class_idx}, ' +
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:108]: `contains`

Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:108]: Replace + by \

File: src/data/datasets/classification/image_classification.py [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+from albumentations import BasicTransform
+from albumentations.core.composition import BaseCompose
+
+from src.registry import DATASETS
+from src.data.datasets.base import ImageDataset
+
+
+@DATASETS.register_class
+class ImageClassificationDataset(ImageDataset):
+    """A generic dataset for multilabel/multiclass image classification task"""
```
Comment (VladVin) [inline src/data/datasets/classification/image_classification.py:15]: Put 2 examples of tables for multiclass and multilabel tasks

File: src/data/datasets/base.py [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+        self.__test_mode = test_mode
+        self.__transform = transform
+        self.__augment = augment
+        self.__input_dtype = input_dtype
+        self.__input_column = input_column
+        self.__grayscale = grayscale
+        self.__transform_targets = transform_targets
+
+        self.__data_folder = Path(data_folder)
+        self.__csv = pd.read_csv(self.data_folder / csv_path)
```
Comment (VladVin) [inline src/data/datasets/base.py:50]: Move to child classes, put `dtype` for each known column.  
`self.__csv = pd.read_csv(self.data_folder / csv_path, dtype={'input_column': str, 'target_column': int or str})`

File: tests/datasets/data/multiclass_test.csv [hunk 23]
```diff
@@ -0,0 +1,6 @@
+image_path,label
```
Comment (VladVin) [inline tests/datasets/data/multiclass_test.csv:1]: Remove files above 100 KB

File: tests/datasets/test_image_classification.py [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+from albumentations.core.composition import Compose
+from albumentations import Resize, HorizontalFlip, VerticalFlip
+
+from src.data.datasets.classification.image_classification import ImageClassificationDataset
+
+
+class TestClassificationMulticlass(unittest.TestCase):
+    def __init__(self, methodName: str = None) -> None:
+        super().__init__(methodName)
+        self.data_folder = os.getcwd() + '/tests/datasets/data/'
```
Comment (VladVin) [inline tests/datasets/test_image_classification.py:15]: Use pathlib and `__file__`: https://stackoverflow.com/a/63313758/3300539

**Review by VladVin — CHANGES_REQUESTED** [review]

File: tests/datasets/test_image_classification.py [hunk 25]
```diff
@@ -0,0 +1,111 @@
+import os
```
Comment (VladVin) [inline tests/datasets/test_image_classification.py:1]: Move tests to `tests/data/datasets/...`

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/data/datasets/base.py [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+            grayscale: If True, image will be read as grayscale otherwise as RGB.
+            test_mode: If True, only image without labels will be returned.
+        """
+        self.__test_mode = test_mode
+        self.__transform = transform
+        self.__augment = augment
+        self.__input_dtype = input_dtype
+        self.__input_column = input_column
+        self.__grayscale = grayscale
+        self.__valid_transform_params = ['image', 'mask', 'bboxes']
```
Comment (VladVin) [inline src/data/datasets/base.py:41]: Remove this error check

**Review by VladVin — APPROVED** [review]

**Pull request merged**